Deploy cockroachdb as k8s statefulset

gprasath opened the pull request:
Signed-off-by: gprasath <giridhara.prasad@cloudbyte.com>

**What this PR does / why we need it**:
- In K8s cluster, deploys cockroachdb application as statefulset with 3 replicas.
- Using new variable in cockroachdb litmusbook for the application deployment type(statefulset/deployment)
- If the deployment is statefulset, verifying if all the replicas are created and running.

gprasath commented:
Sample test run output:
```
giridhara_prasad@strong-eon-153112:~/litmus/apps/cockroachdb/deployers$ kubectl logs -f litmus-cockroachdb-rl98p -n litmus
No config file found; using defaults

PLAY [localhost] ***************************************************************

TASK [Gathering Facts] *********************************************************
ok: [127.0.0.1]

TASK [Generate the litmus result CR to reflect SOT (Start of Test)] ************
changed: [127.0.0.1] => {"changed": true, "checksum": "8dd982d4bbe5573c7bb48a1d359a6d08791b557d", "dest": "./litmus-result.yaml", "gid": 0, "group": "root", "md5sum": "832bc3dc9f59377fa5b25c3fb4775fd1", "mode": "0644", "owner": "root", "size": 419, "src": "/root/.ansible/tmp/ansible-tmp-1535691794.41-4152380677351/source", "state": "file", "uid": 0}

TASK [Apply the litmus result CR] **********************************************
changed: [127.0.0.1] => {"changed": true, "cmd": "kubectl apply -f litmus-result.yaml", "delta": "0:00:01.476314", "end": "2018-08-31 05:03:16.786095", "failed_when_result": false, "rc": 0, "start": "2018-08-31 05:03:15.309781", "stderr": "", "stderr_lines": [], "stdout": "litmusresult \"cockroachdb-deploy\" configured", "stdout_lines": ["litmusresult \"cockroachdb-deploy\" configured"]}

TASK [Check whether the provider storageclass was created.] ********************
changed: [127.0.0.1] => {"changed": true, "cmd": "kubectl get sc openebs-standard", "delta": "0:00:00.930875", "end": "2018-08-31 05:03:17.942405", "rc": 0, "start": "2018-08-31 05:03:17.011530", "stderr": "", "stderr_lines": [], "stdout": "NAME               PROVISIONER\nopenebs-standard   openebs.io/provisioner-iscsi", "stdout_lines": ["NAME               PROVISIONER", "openebs-standard   openebs.io/provisioner-iscsi"]}

TASK [Replace the pvc placeholder with provider] *******************************
changed: [127.0.0.1] => {"changed": true, "msg": "1 replacements made"}

TASK [Replace the storageclass placeholder with provider] **********************
changed: [127.0.0.1] => {"changed": true, "msg": "1 replacements made"}

TASK [Create test specific namespace.] *****************************************
changed: [127.0.0.1] => {"changed": true, "cmd": "source ~/.profile; kubectl create ns cockroachdb", "delta": "0:00:00.936492", "end": "2018-08-31 05:03:19.714972", "rc": 0, "start": "2018-08-31 05:03:18.778480", "stderr": "mesg: ttyname failed: Inappropriate ioctl for device", "stderr_lines": ["mesg: ttyname failed: Inappropriate ioctl for device"], "stdout": "namespace \"cockroachdb\" created", "stdout_lines": ["namespace \"cockroachdb\" created"]}

TASK [Checking the status of test specific namespace.] *************************
changed: [127.0.0.1] => {"attempts": 1, "changed": true, "cmd": "kubectl get ns cockroachdb -o custom-columns=:status.phase --no-headers", "delta": "0:00:00.914066", "end": "2018-08-31 05:03:20.851041", "rc": 0, "start": "2018-08-31 05:03:19.936975", "stderr": "", "stderr_lines": [], "stdout": "Active", "stdout_lines": ["Active"]}

TASK [Deploying cockroachdb statefulset.] **************************************
changed: [127.0.0.1] => {"changed": true, "cmd": "kubectl apply -f cockroachdb-sts.yaml -n cockroachdb", "delta": "0:00:01.076925", "end": "2018-08-31 05:03:22.149306", "rc": 0, "start": "2018-08-31 05:03:21.072381", "stderr": "", "stderr_lines": [], "stdout": "poddisruptionbudget \"cockroachdb-budget\" created\nstatefulset \"cockroachdb\" created", "stdout_lines": ["poddisruptionbudget \"cockroachdb-budget\" created", "statefulset \"cockroachdb\" created"]}

TASK [Deploying cockroachdb service] *******************************************
changed: [127.0.0.1] => {"changed": true, "cmd": "kubectl apply -f cockroachdb-svc.yaml -n cockroachdb", "delta": "0:00:01.055801", "end": "2018-08-31 05:03:23.423389", "rc": 0, "start": "2018-08-31 05:03:22.367588", "stderr": "", "stderr_lines": [], "stdout": "service \"cockroachdb-public\" created\nservice \"cockroachdb\" created", "stdout_lines": ["service \"cockroachdb-public\" created", "service \"cockroachdb\" created"]}

TASK [Check if all the replicas are started] ***********************************
FAILED - RETRYING: Check if all the replicas are started (15 retries left).
changed: [127.0.0.1] => {"attempts": 2, "changed": true, "cmd": "sleep 120; replicas=`kubectl get sts -n cockroachdb -o custom-columns=:spec.replicas`; echo $replicas; ready_replicas=`kubectl get sts -n cockroachdb -o custom-columns=:..readyReplicas`; echo $ready_replicas;", "delta": "0:02:01.121293", "end": "2018-08-31 05:08:26.297433", "rc": 0, "start": "2018-08-31 05:06:25.176140", "stderr": "", "stderr_lines": [], "stdout": "3\n3", "stdout_lines": ["3", "3"]}

TASK [Check if cockroachdb replica pods are running.] **************************
changed: [127.0.0.1] => {"attempts": 1, "changed": true, "cmd": "kubectl get pods -n cockroachdb -o jsonpath='{.items[?(@.metadata.labels.app==\"cockroachdb\")].status.phase}'", "delta": "0:00:00.995480", "end": "2018-08-31 05:08:27.528500", "rc": 0, "start": "2018-08-31 05:08:26.533020", "stderr": "", "stderr_lines": [], "stdout": "Running Running Running", "stdout_lines": ["Running Running Running"]}

TASK [set_fact] ****************************************************************
ok: [127.0.0.1] => {"ansible_facts": {"flag": "Pass"}, "changed": false}

TASK [Generate the litmus result CR to reflect EOT (End of Test)] **************
changed: [127.0.0.1] => {"changed": true, "checksum": "e8b137a2a6e8ada286ec4a7e38f56d7e1de5ab81", "dest": "./litmus-result.yaml", "gid": 0, "group": "root", "md5sum": "2f375a4545a6c95e112d2e62ddfab5ea", "mode": "0644", "owner": "root", "size": 417, "src": "/root/.ansible/tmp/ansible-tmp-1535692107.76-189079277087180/source", "state": "file", "uid": 0}

TASK [Apply the litmus result CR] **********************************************
changed: [127.0.0.1] => {"changed": true, "cmd": "kubectl apply -f litmus-result.yaml", "delta": "0:00:01.076082", "end": "2018-08-31 05:08:31.035973", "failed_when_result": false, "rc": 0, "start": "2018-08-31 05:08:29.959891", "stderr": "", "stderr_lines": [], "stdout": "litmusresult \"cockroachdb-deploy\" configured", "stdout_lines": ["litmusresult \"cockroachdb-deploy\" configured"]}

PLAY RECAP *********************************************************************
127.0.0.1                  : ok=15   changed=13   unreachable=0    failed=0
```

```
giridhara_prasad@strong-eon-153112:~/litmus/apps/cockroachdb/deployers$ kubectl get po -n cockroachdb
NAME                                                              READY     STATUS    RESTARTS   AGE
cockroachdb-0                                                     1/1       Running   0          26m
cockroachdb-1                                                     1/1       Running   0          25m
cockroachdb-2                                                     1/1       Running   0          24m
cockroachdb-datadir-cockroachdb-0-3195640605-ctrl-6986f96dsp77j   2/2       Running   0          26m
cockroachdb-datadir-cockroachdb-0-3195640605-rep-7f9cdc44ccs7n2   1/1       Running   0          26m
cockroachdb-datadir-cockroachdb-0-3195640605-rep-7f9cdc44cgrd6r   1/1       Running   0          26m
cockroachdb-datadir-cockroachdb-0-3195640605-rep-7f9cdc44ctrfjm   1/1       Running   0          26m
cockroachdb-datadir-cockroachdb-1-4179012531-ctrl-655449f7p6kwx   2/2       Running   0          25m
cockroachdb-datadir-cockroachdb-1-4179012531-rep-669fcc6c9gcstt   1/1       Running   0          25m
cockroachdb-datadir-cockroachdb-1-4179012531-rep-669fcc6c9h9wlc   1/1       Running   0          25m
cockroachdb-datadir-cockroachdb-1-4179012531-rep-669fcc6c9t2dvt   1/1       Running   0          25m
cockroachdb-datadir-cockroachdb-2-1316667256-ctrl-6f9597fbtbl97   2/2       Running   0          24m
cockroachdb-datadir-cockroachdb-2-1316667256-rep-dc665f49ddrrpp   1/1       Running   0          24m
cockroachdb-datadir-cockroachdb-2-1316667256-rep-dc665f49dtqpls   1/1       Running   0          24m
cockroachdb-datadir-cockroachdb-2-1316667256-rep-dc665f49dxzx2b   1/1       Running   0          24m
```

dargasudarshan commented on apps/cockroachdb/deployers/cockroachdb-installer.yaml:
```diff
@@ -0,0 +1,122 @@
+            replicas=`kubectl get sts -n {{ namespace }} -o custom-columns=:spec.replicas`; 
+            echo $replicas;
+            ready_replicas=`kubectl get sts -n {{ namespace }} -o custom-columns=:..readyReplicas`; 
+            echo $ready_replicas;
```
@gprasath Can we split the ready replica task separately to avoid sleep and include until condition to get the statefulset replicas ready status?

gprasath replied:
Thanks @dargasudarshan . Now, checking the tasks with return code and validating the condition.

gprasath commented:
Sample test run output after modification:

```
giridhara_prasad@strong-eon-153112:~/litmus/apps/cockroachdb/deployers$ kubectl logs -f litmus-cockroachdb-8jqd6 -n litmus
No config file found; using defaults

PLAY [localhost] ***************************************************************

TASK [Gathering Facts] *********************************************************
ok: [127.0.0.1]

TASK [Generate the litmus result CR to reflect SOT (Start of Test)] ************
changed: [127.0.0.1] => {"changed": true, "checksum": "8dd982d4bbe5573c7bb48a1d359a6d08791b557d", "dest": "./litmus-result.yaml", "gid": 0, "group": "root", "md5sum": "832bc3dc9f59377fa5b25c3fb4775fd1", "mode": "0644", "owner": "root", "size": 419, "src": "/root/.ansible/tmp/ansible-tmp-1535697290.97-254283223927201/source", "state": "file", "uid": 0}

TASK [Apply the litmus result CR] **********************************************
changed: [127.0.0.1] => {"changed": true, "cmd": "kubectl apply -f litmus-result.yaml", "delta": "0:00:01.349612", "end": "2018-08-31 06:34:53.188477", "failed_when_result": false, "rc": 0, "start": "2018-08-31 06:34:51.838865", "stderr": "", "stderr_lines": [], "stdout": "litmusresult \"cockroachdb-deploy\" configured", "stdout_lines": ["litmusresult \"cockroachdb-deploy\" configured"]}

TASK [Check whether the provider storageclass was created.] ********************
changed: [127.0.0.1] => {"changed": true, "cmd": "kubectl get sc openebs-standard", "delta": "0:00:00.919359", "end": "2018-08-31 06:34:54.332747", "rc": 0, "start": "2018-08-31 06:34:53.413388", "stderr": "", "stderr_lines": [], "stdout": "NAME               PROVISIONER\nopenebs-standard   openebs.io/provisioner-iscsi", "stdout_lines": ["NAME               PROVISIONER", "openebs-standard   openebs.io/provisioner-iscsi"]}

TASK [Replace the pvc placeholder with provider] *******************************
changed: [127.0.0.1] => {"changed": true, "msg": "1 replacements made"}

TASK [Replace the storageclass placeholder with provider] **********************
changed: [127.0.0.1] => {"changed": true, "msg": "1 replacements made"}

TASK [Create test specific namespace.] *****************************************
changed: [127.0.0.1] => {"changed": true, "cmd": "source ~/.profile; kubectl create ns cockroachdb", "delta": "0:00:00.906638", "end": "2018-08-31 06:34:56.083851", "rc": 0, "start": "2018-08-31 06:34:55.177213", "stderr": "mesg: ttyname failed: Inappropriate ioctl for device", "stderr_lines": ["mesg: ttyname failed: Inappropriate ioctl for device"], "stdout": "namespace \"cockroachdb\" created", "stdout_lines": ["namespace \"cockroachdb\" created"]}

TASK [Checking the status of test specific namespace.] *************************
changed: [127.0.0.1] => {"attempts": 1, "changed": true, "cmd": "kubectl get ns cockroachdb -o custom-columns=:status.phase --no-headers", "delta": "0:00:00.926842", "end": "2018-08-31 06:34:57.225416", "rc": 0, "start": "2018-08-31 06:34:56.298574", "stderr": "", "stderr_lines": [], "stdout": "Active", "stdout_lines": ["Active"]}

TASK [Deploying cockroachdb statefulset.] **************************************
changed: [127.0.0.1] => {"changed": true, "cmd": "kubectl apply -f cockroachdb-sts.yaml -n cockroachdb", "delta": "0:00:01.056403", "end": "2018-08-31 06:34:58.500112", "rc": 0, "start": "2018-08-31 06:34:57.443709", "stderr": "", "stderr_lines": [], "stdout": "poddisruptionbudget \"cockroachdb-budget\" created\nstatefulset \"cockroachdb\" created", "stdout_lines": ["poddisruptionbudget \"cockroachdb-budget\" created", "statefulset \"cockroachdb\" created"]}

TASK [Deploying cockroachdb service] *******************************************
changed: [127.0.0.1] => {"changed": true, "cmd": "kubectl apply -f cockroachdb-svc.yaml -n cockroachdb", "delta": "0:00:01.085063", "end": "2018-08-31 06:34:59.797285", "rc": 0, "start": "2018-08-31 06:34:58.712222", "stderr": "", "stderr_lines": [], "stdout": "service \"cockroachdb-public\" created\nservice \"cockroachdb\" created", "stdout_lines": ["service \"cockroachdb-public\" created", "service \"cockroachdb\" created"]}

TASK [Obtain the number of replicas.] ******************************************
changed: [127.0.0.1] => {"attempts": 1, "changed": true, "cmd": "replicas=`kubectl get sts -n cockroachdb -o custom-columns=:spec.replicas`; echo $replicas;", "delta": "0:00:00.920463", "end": "2018-08-31 06:35:00.952973", "rc": 0, "start": "2018-08-31 06:35:00.032510", "stderr": "", "stderr_lines": [], "stdout": "3", "stdout_lines": ["3"]}

TASK [Obtain the ready replica count and compare with the replica count.] ******
FAILED - RETRYING: Obtain the ready replica count and compare with the replica count. (15 retries left).
FAILED - RETRYING: Obtain the ready replica count and compare with the replica count. (14 retries left).
FAILED - RETRYING: Obtain the ready replica count and compare with the replica count. (13 retries left).
changed: [127.0.0.1] => {"attempts": 4, "changed": true, "cmd": "ready_replicas=`kubectl get sts -n cockroachdb -o custom-columns=:..readyReplicas`; echo $ready_replicas;", "delta": "0:00:00.957142", "end": "2018-08-31 06:38:05.684393", "rc": 0, "start": "2018-08-31 06:38:04.727251", "stderr": "", "stderr_lines": [], "stdout": "3", "stdout_lines": ["3"]}

TASK [Check if cockroachdb replica pods are running.] **************************
changed: [127.0.0.1] => {"attempts": 1, "changed": true, "cmd": "kubectl get pods -n cockroachdb -o jsonpath='{.items[?(@.metadata.labels.app==\"cockroachdb\")].status.phase}'", "delta": "0:00:00.995618", "end": "2018-08-31 06:38:06.919124", "rc": 0, "start": "2018-08-31 06:38:05.923506", "stderr": "", "stderr_lines": [], "stdout": "Running Running Running", "stdout_lines": ["Running Running Running"]}

TASK [set_fact] ****************************************************************
ok: [127.0.0.1] => {"ansible_facts": {"flag": "Pass"}, "changed": false}

TASK [Generate the litmus result CR to reflect EOT (End of Test)] **************
changed: [127.0.0.1] => {"changed": true, "checksum": "e8b137a2a6e8ada286ec4a7e38f56d7e1de5ab81", "dest": "./litmus-result.yaml", "gid": 0, "group": "root", "md5sum": "2f375a4545a6c95e112d2e62ddfab5ea", "mode": "0644", "owner": "root", "size": 417, "src": "/root/.ansible/tmp/ansible-tmp-1535697487.14-232923374826749/source", "state": "file", "uid": 0}

TASK [Apply the litmus result CR] **********************************************
changed: [127.0.0.1] => {"changed": true, "cmd": "kubectl apply -f litmus-result.yaml", "delta": "0:00:01.088976", "end": "2018-08-31 06:38:10.388052", "failed_when_result": false, "rc": 0, "start": "2018-08-31 06:38:09.299076", "stderr": "", "stderr_lines": [], "stdout": "litmusresult \"cockroachdb-deploy\" configured", "stdout_lines": ["litmusresult \"cockroachdb-deploy\" configured"]}

PLAY RECAP *********************************************************************
127.0.0.1                  : ok=16   changed=14   unreachable=0    failed=0
```

ksatchit commented on apps/cockroachdb/deployers/cockroachdb-installer.yaml:
```diff
@@ -0,0 +1,129 @@
+
+        - block:
+
+            - name: Obtain the number of replicas.
```
There maybe other statefulsets on the cluster too. Can we filter by labels (we could use common labels for all type of deployments)

gprasath replied:
Included. Thanks @ksatchit
